include/seastar/core/shared_mutex.hh

```diff
@@ -193,7 +193,33 @@ with_shared(shared_mutex& sm, Func&& func) noexcept {
             });
         });
     } catch (...) {
-        return current_exception_as_future();
+        return current_exception_as_future<>();
+    }
+}
+
+template <typename Func>
+SEASTAR_CONCEPT(
+    requires (std::invocable<Func> && !std::is_nothrow_move_constructible_v<Func> && !std::same_as<futurize_t<std::invoke_result_t<Func>>, future<>>)
+    inline
+    futurize_t<std::invoke_result_t<Func>>
+)
```

Comment:
You can use `futurize<T>::current_exception_as_future()` to avoid the duplication.

Reply:
sounds like a good idea, I'll take a stab at it